auth/role.go

```diff
@@ -20,15 +20,33 @@ import (
 
 /** A group that users can belong to, with associated channel permissions. */
 type roleImpl struct {
-	Name_             string      `json:"name,omitempty"`
-	ExplicitChannels_ ch.TimedSet `json:"admin_channels,omitempty"`
-	Channels_         ch.TimedSet `json:"all_channels"`
-	Sequence_         uint64      `json:"sequence"`
-	PreviousChannels_ ch.TimedSet `json:"previous_channels,omitempty"`
+	Name_             string                  `json:"name,omitempty"`
+	ExplicitChannels_ ch.TimedSet             `json:"admin_channels,omitempty"`
+	Channels_         ch.TimedSet             `json:"all_channels"`
+	Sequence_         uint64                  `json:"sequence"`
+	PreviousChannels_ *PreviousChannelsOrRole `json:"previous_channels,omitempty"`
+	ChannelHistory_   ChannelOrRoleHistory    `json:"channel_history,omitempty"`
 	vbNo              *uint16
 	cas               uint64
 }
 
+type PreviousChannelsOrRole struct {
```

Comment:
I think it would be better to have a more generic name for this struct, instead of PreviousChannelsOrRole.  InvalidatedTimedSet seems like one possibility.

Can we also add a description for this struct, especially to describe what invalSeq represents, compared to the sequences that are included in the timed set?

Reply:
Again, will add comment but to clarify here>
invalSeq: The sequence the change was triggered by as obtained from MarkPrincipalsChanged, Later used for triggeredBy
TimedSet: Just contains previous data

Reply:
Removed previous